frontends/p4/validateP4runtimeTranslationAnnotation.h

```diff
@@ -0,0 +1,56 @@
+/*
+Copyright 2019 MNK Consulting, LLC.
+  http://mnkcg.com
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+#ifndef _P4_VALIDATEP4TRANSLATIONANNOTATIONS_H_
+#define _P4_VALIDATEP4TRANSLATIONANNOTATIONS_H_
+
+#include "lib/error.h"
+#include "ir/ir.h"
+#include "frontends/p4/typeMap.h"
+
+namespace P4 {
+
+/**
+ * Checks that p4runtime annotations.
+ */
+class ValidateP4runtimeTranslationAnnotations final : public Inspector {
```

Comment:
The way I see it this pass is superfluous. It seems strictly less useful that `PARSE_PAIR("p4runtime_translation", StringLiteral, Constant)`

Reply:
```
In file included from /tmp/p4c/build/frontends/unified_frontend_sources_4.cpp:7:0:
/tmp/p4c/frontends/p4/parseAnnotations.cpp:49:17: error: macro "PARSE_PAIR" passed 3 arguments, but takes just 2
         Constant),
                 ^
/tmp/p4c/frontends/p4/parseAnnotations.cpp:51:34: error: macro "PARSE_TRIPLE" passed 3 arguments, but takes just 2
                          Constant),
                                  ^
```

Reply:
It seems `PARSE_PAIR` cannot handle mixed types. Maybe it's worth figuring out if this can be improved. Maybe `PARSE_PAIR("p4runtime_translation", Expression)` will work. It's not as food because it is not as strict, but it may parse the annotation in a helpful way and your pass currently doesn't appear to validate the types at the moment anyway.

Reply:
This was the first one I tried, but it fails at run-time.

```
hemant@ubuntu:/tmp/p4c/build$ p4/testdata/p4_16_samples/p416-type-use3.p4.test -f
./p4test --pp ./tmppVngj8/p416-type-use3.p4 --dump ./tmppVngj8 --top4 FrontEndLast,FrontEndDump,MidEndLast --testJson --arch v1model --p4runtime-files ./tmppVngj8/p416-type-use3.p4.p4info.txt /tmp/p4c/testdata/p4_16_samples/p416-type-use3.p4
Error compiling
p416-type-use3.p4(33):syntax error, unexpected ",", expecting END_ANNOTATION
@p4runtime_translation("p4.org/psa/v1/EthT_t", 48)
                                             ^
```
It won't take much to add validation of the types in my new pass.  Sorry, for the delay - I am at the P4 Hackathon today.

Reply:
`PARSE_PAIR("p4runtime_translation", Expression)` is supposed to work. Unfortunately there is no example / test in the p4c repo, but I can see it being used successfully in the Barefoot proprietary compiler backend. Maybe the annotation parser doesn't like the backslashes? Any idea @liujed ?

Reply:
I'm not sure what is wrong. That really should work. @hesingh, if you make your code available and tell me how to reproduce, I can take a look.

Reply:
@antoninbas I see at least two problems with PARSE.  First, I added a backslash before each forward slash and then the error cursor moved to the comma.  The second issue is that the PARSE thinks the Expression ends when the second double-quotes is read. 

Regarding the new pass I added, it's is doing validation.  See

https://github.com/p4lang/p4c/pull/1736/files#diff-c0775e19ed0e36ad9a91c029875df5e1R44

If additional validation is required such as to verify the StringLiternal (does "p4.org/psa/v1/" 
exist in string), does a comma exist as second element, and the last element is a Constant, it's trivial to add. 

@liujed The workspace information is included below.

https://github.com/familysingh/p4c

1. Please edit `p4c/frontend/CMakeLists.txt` to remove
`validateP4runtimeTranslationAnnotation.h` from it.
2. In p4c/build/ directory, invoke "cmake .."
3. Add this line of code to `parseAnnotations.cpp`
`PARSE(IR::Annotation::p4runtimeTranslationAnnotation, Expression),`
4. Build p4c
5. Run test. 
```p4c/build$p4/testdata/p4_16_samples/p416-type-use3.p4.test 

hemant@ubuntu:/tmp/p4c/build$ p4/testdata/p4_16_samples/p416-type-use3.p4.test
./p4test --pp ./tmpOJ_e9P/p416-type-use3.p4 --dump ./tmpOJ_e9P --top4 FrontEndLast,FrontEndDump,MidEndLast --testJson --arch v1model --p4runtime-files ./tmpOJ_e9P/p416-type-use3.p4.p4info.txt /tmp/p4c/testdata/p4_16_samples/p416-type-use3.p4
Error compiling
p416-type-use3.p4(33):syntax error, unexpected ",", expecting END_ANNOTATION
@p4runtime_translation("p4.org/psa/v1/EthT_t", 48)
                                             ^

hemant@ubuntu:/tmp/p4c/build$ 
```
6. Example of the Annotation exists in p416-type-use3.p4.test.  Look for 

`@p4runtime_translation("p4.org/psa/v1/EthT_t", 48)`

Reply:
Looks like you misread Antonin's suggestion—I believe it should be `PARSE_PAIR(...)` instead of `PARSE(...)`.

Reply:
Please take a log at the logs I have added above.  Even `PARSE_PAIR()` fails.

I am including the same failed log again.

This was the p4c code used:

```PARSE_PAIR(IR::Annotation::p4runtimeTranslationAnnotation, Expression),```

```
hemant@ubuntu:/tmp/p4c/build$ p4/testdata/p4_16_samples/p416-type-use3.p4.test 
./p4test --pp ./tmpsA6pJb/p416-type-use3.p4 --dump ./tmpsA6pJb --top4 FrontEndLast,FrontEndDump,MidEndLast --testJson --arch v1model --p4runtime-files ./tmpsA6pJb/p416-type-use3.p4.p4info.txt /tmp/p4c/testdata/p4_16_samples/p416-type-use3.p4
Error compiling
p416-type-use3.p4(33): [--Werror=invalid] error: @p4runtime_translation: Invalid @p4runtime_translation: annotation must have exactly 1 argument
@p4runtime_translation("p4.org/psa/v1/EthT_t", 48)
^
```

Reply:
That message would be coming from the validation you added: https://github.com/familysingh/p4c/blob/master/frontends/p4/validateP4runtimeTranslationAnnotation.h#L38-L40. If you are parsing a pair of expressions, there will be two expressions, not just one.

Reply:
You are correct.  Sorry, I thought I had disabled my pass to validate Annotation in my build.  I have used PARSE_PAIR with one test and it works - thanks.